Add support for nullable analysis in interpolated string handler constructors

333fred opened the pull request:
Implements most of https://github.com/dotnet/roslyn/issues/54583. Arguments with slots are treated like we do receivers: as though the previous argument value has already been loaded, and therefore postcondition attributes do not flow back to the original argument position (the effect of this can be seen in StringInterpolation_14 and _20), but future references to those locations will see the effects of the attribute.

I additionally looked at updating region analysis as part of this PR to see if I could share data between these passes, but I don't believe we can do much sharing here. There's too much nullable-specific with finding the current state of the expression being used for the mapping to work.

RikkiGibson commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -64270,6 +64270,438 @@ public CustomHandler(int literalLength, int formattedCount" + (validityParameter
+
+M(out string? s1, $"""");
+
+void M(out string? t1, [InterpolatedStringHandlerArgument(""t1"")] CustomHandler c) => t1 = null;
```
I feel silly for asking at this stage but.. I am wondering what is supposed to happen when the user sees this in intellisense? Are most APIs going to be overloaded on 'string' and 'CustomHandler' in practice so the user will see the string overload and say "ok, I pretty much know what to provide here"?

333fred replied:
Likely most will overload, but nothing special will be shown for custom handlers.

RikkiGibson commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3261,7 +3272,18 @@ private void VisitObjectElementInitializer(int containingSlot, BoundAssignmentOp
+                node.DefaultArguments,
+                node.Expanded,
+                node.ImplicitReceiverOpt,
+                receiverState: NullableFlowState.NotNull,
```
It feels like whenever we might have a not-null receiver, we pass a NotNull flow state here. Is it necessary for this to be its own parameter?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -3011,7 +3010,7 @@ private BoundExpression BindArgumentExpression(BindingDiagnosticBag diagnostics,
                     TypeWithAnnotations parameterTypeWithAnnotations = GetCorrespondingParameterTypeWithAnnotations(ref result, parameters, arg);
                     reportUnsafeIfNeeded(methodResult, diagnostics, argument, parameterTypeWithAnnotations);
-                    arguments[arg] = BindInterpolatedStringHandlerInMemberCall(argument, arguments, parameters, ref result, arg, receiverType, receiverEscapeScope, diagnostics);
+                    arguments[arg] = BindInterpolatedStringHandlerInMemberCall(argument, arguments, parameters, ref result, arg, receiver, methodResult.LeastOverriddenMember, diagnostics);
```
>LeastOverriddenMember [](http://example.com/codeflow?start=152&length=21)

It is not obvious why we have to pass the ```LeastOverriddenMember``` here. If it is a requirement specific to binding an interpolated string handler, perhaps ```BindInterpolatedStringHandlerInMemberCall``` should do that itself rather than relying on the caller to supply the right value.

333fred replied:
I'll reduce the symbol to just a bool for whether it requires an instance receiver.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -962,33 +962,40 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
                 switch (argumentIndex)
                 {
-                    case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter:
+                    case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter when receiver != null:
```
>when receiver != null [](http://example.com/codeflow?start=86&length=21)

It looks like the switch above over the same ```argumentIndex``` assumes the ```receiver``` is not null for ```case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter```. Here we are guarding against null in the same situation. Is there a reason for the inconsistency?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -962,33 +962,40 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
+                    case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter when receiver != null:
                         placeholderSyntax = unconvertedString.Syntax;
-                        valSafeToEscapeScope = receiverEscapeScope;
+                        valSafeToEscapeScope = calledMember.RequiresInstanceReceiver()
```
>valSafeToEscapeScope [](http://example.com/codeflow?start=24&length=20)

Is this change related to the stated goal of the PR, i.e. support for nullable analysis? Is there a change in behavior here?

333fred replied:
> Is this change related to the stated goal of the PR, i.e. support for nullable analysis? Is there a change in behavior here?

Yes, it's related, though this itself shouldn't cause a behavior change. I needed to pass through additional context from the receiver for the correct syntax node to report nullability warnings, and whether the receiver has a null suppression on it. I decided to simplify the existing receiver data to just pass the receiver itself.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -962,33 +962,40 @@ private ImmutableArray<BoundExpression> BindInterpolatedStringParts(BoundUnconve
                 {
-                    case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter:
+                    case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter when receiver != null:
                         placeholderSyntax = unconvertedString.Syntax;
```
>placeholderSyntax = unconvertedString.Syntax; [](http://example.com/codeflow?start=24&length=45)

Is this assignment necessary.

AlekseyTs commented:
@333fred It looks like there are legitimate test failures.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4199,6 +4248,11 @@ private void LearnFromNullTest(BoundExpression expression, ref LocalState state)
 
+            if (slot < 0 && expressionWithoutConversion.Kind == BoundKind.InterpolatedStringArgumentPlaceholder)
+            {
+                _ = TryGetPlaceholderSlot(expressionWithoutConversion, out slot);
```
>_ = TryGetPlaceholderSlot(expressionWithoutConversion, out slot); [](http://example.com/codeflow?start=16&length=65)

Should MakeSlot take care of this side check instead?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -4128,6 +4169,14 @@ private void GetSlotsToMarkAsNotNullable(BoundExpression operand, ArrayBuilder<i
                             break;
 
+                        case BoundKind.InterpolatedStringArgumentPlaceholder:
+                            if (TryGetPlaceholderSlot(operand, out slot) && PossiblyNullableType(operand.Type))
```
>if (TryGetPlaceholderSlot(operand, out slot) && PossiblyNullableType(operand.Type)) [](http://example.com/codeflow?start=28&length=83)

It feels like we wouldn't need this special case if MakeSlot took care of the side lookup.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5569,16 +5630,64 @@ void markMembersAsNotNull(int receiverSlot, TypeSymbol type, string memberName,
+
+                if (possibleInterpolationData is ({ ArgumentPlaceholders.Length: > 1 }
+                                                  or { ArgumentPlaceholders.Length: 1, HasTrailingHandlerValidityParameter: false })
+                                                 and var interpolationData)
```
If I am correctly interpreting the intent here, I think the following condition would be easier to follow:
```possibleInterpolationData is not null && possibleInterpolationData.ArgumentPlaceholders.Length > (possibleInterpolationData.HasTrailingHandlerValidityParameter ? 1 : 0)```. Then the block can keep using ```possibleInterpolationData```.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5569,16 +5630,64 @@ void markMembersAsNotNull(int receiverSlot, TypeSymbol type, string memberName,
+                                break;
+                            case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter:
+                                int slot = getInterpolatedStringHandlerPlaceholderSlot(placeholder, receiver);
+                                State[slot] = receiverState;
```
>State[slot] = receiverState; [](http://example.com/codeflow?start=32&length=28)

It is not clear to me why overwriting the state for the real receiver's slot is the right thing to do.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5569,16 +5630,64 @@ void markMembersAsNotNull(int receiverSlot, TypeSymbol type, string memberName,
+                                if (resultsBuilder.Count > placeholder.ArgumentIndex)
+                                {
+                                    slot = getInterpolatedStringHandlerPlaceholderSlot(placeholder, arguments[placeholder.ArgumentIndex]);
+                                    State[slot] = resultsBuilder[placeholder.ArgumentIndex].VisitResult.RValueType.State;
```
>State[slot] [](http://example.com/codeflow?start=36&length=11)

Same comment here.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3341,6 +3363,25 @@ private int GetOrCreatePlaceholderSlot(object identifier, TypeWithAnnotations ty
             return GetOrCreateSlot(placeholder, forceSlotEvenIfEmpty: true);
         }
 
+        private bool TryGetPlaceholderSlot(BoundExpression node, out int slot)
```
>BoundExpression [](http://example.com/codeflow?start=43&length=15)

It looks like this function is only used for BoundInterpolatedStringArgumentPlaceholder. I think we can benefit from making the parameter type stronger.

333fred replied:
This function is being removed.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3341,6 +3363,25 @@ private int GetOrCreatePlaceholderSlot(object identifier, TypeWithAnnotations ty
 
+        private bool TryGetPlaceholderSlot(BoundExpression node, out int slot)
+        {
+            if (_placeholderLocalsOpt?.TryGetValue(node, out var placeholderLocal) == true)
```
>_placeholderLocalsOpt [](http://example.com/codeflow?start=16&length=21)

I think the implementation can be simpler if we change ```getInterpolatedStringHandlerPlaceholderSlot``` to always add an entry into ```_interpolatedStringArgumentPlaceholderAliases```. This way we wouldn't need to look in two places.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3341,6 +3363,25 @@ private int GetOrCreatePlaceholderSlot(object identifier, TypeWithAnnotations ty
             return GetOrCreateSlot(placeholder, forceSlotEvenIfEmpty: true);
         }
 
+        private bool TryGetPlaceholderSlot(BoundExpression node, out int slot)
```
>TryGetPlaceholderSlot [](http://example.com/codeflow?start=21&length=21)

The name feels somewhat confusing given the name of the previous methods. It feels like that apply different meaning to the "Placeholder" word. Perhaps this name should be more specific.

AlekseyTs commented:
> Arguments with slots are treated like we do receivers: as though the previous argument value has already been loaded, and therefore postcondition attributes do not flow back to the original argument position

Is this something that should be approved by LDM? Also, do we plan to update nullable specification?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -9597,8 +9708,12 @@ private TypeWithState InferResultNullabilityOfBinaryLogicalOperator(BoundExpress
             }
+            else
+            {
+                receiverOpt = null;
```
>receiverOpt = null; [](http://example.com/codeflow?start=16&length=19)

It is not clear to me why we start pretending that there is no receiver. Perhaps, this allows us to get what we want (BTW, what is it) from ```VisitArgumentsEvaluate``` call below today. But we might get in trouble if we start using this value for something else.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5547,13 +5607,14 @@ void markMembersAsNotNull(int receiverSlot, TypeSymbol type, string memberName,
-            bool expanded)
+            bool expanded,
+            BoundExpression? receiver,
+            NullableFlowState receiverState)
```
>receiverState [](http://example.com/codeflow?start=30&length=13)

It looks like we always pass one of the two values:
- default
- NotNull.

More over, default is passed only when we explicitly pass ```nulll``` for the receiver. What is the advantage of using ```default``` state vs. NotNull state in cases like that? Perhaps we don't need to pass the ```receiverState``` explicitly at all?

AlekseyTs commented:
Done with review pass (commit 2)

333fred commented:
Converting this to a draft PR while I'm out on vacation.

333fred commented:
> Also, do we plan to update nullable specification?

I'll update the specification with the limitations after this is merged.

---
In reply to: [974418424](https://github.com/dotnet/roslyn/pull/57780#issuecomment-974418424) [](http://example.com/codeflow?ancestors=974418424,974418424)

333fred commented:
@AlekseyTs addressed your feedback and did a bit of rearranging. I did have to merge from `main` to pull in the placeholder work you did, so the diff isn't the cleanest between 2 and 9, unfortunately.

---
In reply to: [974438542](https://github.com/dotnet/roslyn/pull/57780#issuecomment-974438542) [](http://example.com/codeflow?ancestors=974438542,974438542)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -245,5 +245,14 @@ static void pushLeftNodes(BoundBinaryOperator binary, ArrayBuilder<BoundBinaryOp
+                BoundBinaryOperator { InterpolatedStringHandlerData: { } d } => d,
+                BoundInterpolatedString { InterpolationData: { } d } => d,
+                BoundBinaryOperator or BoundInterpolatedString when !throwOnMissing => default,
+                _ => throw ExceptionUtilities.UnexpectedValue(e.Kind),
```
>throw ExceptionUtilities.UnexpectedValue(e.Kind), [](http://example.com/codeflow?start=21&length=49)

```!throwOnMissing ? default : throw ExceptionUtilities.UnexpectedValue(e.Kind)``` and get rid of the previous item in the switch?

AlekseyTs replied:
>e.Kind [](http://example.com/codeflow?start=62&length=6)

Complaining about unexpected node for ```BoundBinaryOperator``` or ```BoundInterpolatedString``` might be confusing.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5578,9 +5594,11 @@ void markMembersAsNotNull(int receiverSlot, TypeSymbol type, string memberName,
                 // we disable nullable warnings on default arguments
                 _disableDiagnostics = defaultArguments[i] || previousDisableDiagnostics;
-                resultsBuilder.Add(VisitArgumentEvaluate(arguments[i], GetRefKind(refKindsOpt, i), parameterAnnotations));
+                resultsBuilder.Add(VisitArgumentEvaluate(argument, GetRefKind(refKindsOpt, i), parameterAnnotations));
```
>argument [](http://example.com/codeflow?start=57&length=8)

It looks like we cold keep the original code for the body of this function.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5653,9 +5671,12 @@ private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefK
             bool extensionMethodThisArgument)
         {
             Debug.Assert(!this.IsConditionalState);
+            Debug.Assert(conversionResultsBuilder == null || conversionResultsBuilder.Count == argumentIndex);
```
>conversionResultsBuilder.Count == argumentIndex) [](http://example.com/codeflow?start=61&length=48)

Could we infer the index from the count then?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -6968,10 +6992,12 @@ void reportBadDelegateParameter(BindingDiagnosticBag bag, MethodSymbol sourceInv
         {
             Debug.Assert(!trackMembers || !IsConditionalState);
             Debug.Assert(conversionOperand != null);
+            Debug.Assert(interpolatedStringArgumentInformation is not { } info || info.PreviousArgumentConverionResults.Count == info.ArgumentIndex);
```
>info.PreviousArgumentConverionResults.Count == info.ArgumentIndex [](http://example.com/codeflow?start=82&length=65)

Could we infer the index from the count then?

AlekseyTs commented:
Done with review pass (commit 9), it looks like there are merge conflicts to resolve.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [987109770](https://github.com/dotnet/roslyn/pull/57780#issuecomment-987109770) [](http://example.com/codeflow?ancestors=987109770,987109770)

AlekseyTs commented:
@333fred It looks like there are legitimate test failures

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -9914,8 +9993,6 @@ private static bool IsNullabilityMismatch(TypeSymbol type1, TypeSymbol type2)
         {
-            // These placeholders don't yet follow proper placeholder discipline
-            AssertPlaceholderAllowedWithoutRegistration(node);
             SetNotNullResult(node);
```
>SetNotNullResult [](http://example.com/codeflow?start=12&length=16)

VisitPlaceholderWithReplacement?

333fred replied:
I am unsure how this got reverted, must have been something with the merge?

AlekseyTs commented:
Done with review pass (commit 11).

333fred commented:
@AlekseyTs addressed feedback, thanks.

---
In reply to: [987441294](https://github.com/dotnet/roslyn/pull/57780#issuecomment-987441294) [](http://example.com/codeflow?ancestors=987441294,987441294)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -7481,6 +7503,62 @@ NullableFlowState visitNestedTargetTypedConstructs()
+                    {
+                        if (placeholder.ArgumentIndex >= 0)
+                        {
+                            _resultForPlaceholdersOpt.Remove(placeholder);
```
>_resultForPlaceholdersOpt.Remove [](http://example.com/codeflow?start=28&length=32)

```RemovePlaceholderReplacement```? And remove the assert before the loop?

AlekseyTs reviewed (dismissed):
LGTM (commit 12), modulo the small refactoring suggestion around an explicit use of ```_resultForPlaceholdersOpt``` field.

333fred commented:
@dotnet/roslyn-compiler @RikkiGibson for a second review.

333fred commented:
@AlekseyTs, a test I added in this PR revealed a bug in the IOperation implementation, please take a look at the latest commit.

333fred commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -1545,7 +1545,6 @@ public override void VisitFlowCapture(IFlowCaptureOperation operation)
                     break;
 
                 case OperationKind.OmittedArgument:
-                case OperationKind.DeclarationExpression:
```
Because, in the case of interpolated string handlers, the declaration expression is actually captured. See the graph in the test case I added, which would fail here because the declaration expression goes through here.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1839,10 +1839,21 @@ private void SpillEvalStack()
+                        // node underlying the declaration. Currently, this should only happen for `out` variable declarations,
+                        // which can only be field or local references. If we ever encounter a scenario where tuples get here,
+                        // we will need to update the logic to ensure we aren't erasing information.
+                        operationOpt = OperationCloner.CloneOperation(declarationExpression.Expression);
```
>CloneOperation [](http://example.com/codeflow?start=55&length=14)

Is the clone necessary?

333fred replied:
Yes. The expression already has a `Parent` and we can't reset it.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1802,7 +1802,7 @@ private int GetNextCaptureId(RegionBuilder owner)
         }
 
-        private void SpillEvalStack()
+        private void SpillEvalStack(bool spillDeclarationExpressions = false)
```
It feels like we might spill too much. I.e. spill `DeclarationExpression`s that do not belong to the immediate explicit call. We probably should limit this exception by ```_currentInterpolatedStringHandlerArgumentContext.StartingStackDepth```, but then we also might need to drop the optimization based on ```_startSpillingAt```.

AlekseyTs replied:
On the second thought, this optimization will come at the cost of figuring out the right spill region for the spilling "out of order". Not worth the risk and complexity, I think.

AlekseyTs commented:
Done with review pass (commit 15)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6401,7 +6412,7 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
             // handler semantics), and then evaluate to the handler flow capture temp.
 
-            SpillEvalStack();
+            SpillEvalStack(spillDeclarationExpressions: true);
```
>spillDeclarationExpressions: true [](http://example.com/codeflow?start=27&length=33)

This doesn't feel sufficient. This parameter will make the difference only for entries that haven't been spilled already. And there is no guarantee that that will be the case. For example, if an argument following a ```DeclarationExpression``` has a control flow, it will case spilling for the ```DeclarationExpression``` and after that this call won't do anything about it. It should be easy to come up with a test case that exposes this issue.
Some options to address this that come to mind:
- Unconditionally unwrap ```DeclarationExpression```s during spilling. If we can run into tuple-like ```DeclarationExpression```s, perhaps we can skip those.
- Preemptively unwrap ```DeclarationExpression```s when we push arguments to the stack.
- Handle spilling of ```DeclarationExpression```out of order on demand once we realize that we need to use it multiple times due to the interpolation.
We can discuss our options in more details offline.

333fred commented:
@AlekseyTs I implemented the change we talked about offline. Please take another look when you get the chance.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1802,7 +1802,7 @@ private int GetNextCaptureId(RegionBuilder owner)
         }
 
-        private void SpillEvalStack()
+        private void SpillEvalStack(bool spillDeclarationExpressions = false)
```
>bool spillDeclarationExpressions = false [](http://example.com/codeflow?start=36&length=40)

Is the parameter still necessary.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:
```diff
@@ -5809,6 +5809,280 @@ public CustomHandler(int literalLength, int formattedCount, out int i) : this(li
+{
+    public static void M1(object o1, object o2)
+    /*<bind>*/{
+        C.M(out int i, o1 ?? o2, $""literal{1}"");
```
>); [](http://example.com/codeflow?start=48&length=2)

Consider adding a test with declaration expression between two interpolations and also a test demonstrating that declaration expressions at the end survive.

AlekseyTs commented:
>         private static readonly Func<IArgumentOperation, IOperation> UnwrapArgument = UnwrapArgumentDoNotCaptureDirectly;

Is this delegate used?

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:2092 in 733946f. [](commit_id = 733946f06683acf102fa7505efd453721f59603f, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2036,17 +2047,45 @@ private void VisitAndPushArguments(ImmutableArray<IArgumentOperation> arguments,
+                // this check is more complexity than this scenario needs.
+                if (i < lastIndexForSpilling)
+                {
+                    SpillEvalStack(spillDeclarationExpressions: true);
```
>SpillEvalStack(spillDeclarationExpressions: true); [](http://example.com/codeflow?start=20&length=50)

Instead of doing this, I would simply conditionally unwrap declaration expressions before calling ```VisitRequired``` above.

AlekseyTs replied:
Leaving the logic in SpillEvalStack feels fragile because it will continue to be a trap for user that might think that passing ```true``` actually provides guarantees on its own. Another shortcoming is that this call is going to unwrap declaration expressions that do not belong to this argument list. For example, arguments of an enclosing call. Please add a test case for this scenario.

AlekseyTs commented:
Done with review pass (commit 16).

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [991353417](https://github.com/dotnet/roslyn/pull/57780#issuecomment-991353417) [](http://example.com/codeflow?ancestors=991353417,991353417)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2036,25 +2036,47 @@ private void VisitAndPushArguments(ImmutableArray<IArgumentOperation> arguments,
+                // have unexpected affects on consumers. So we limit the spilling to those indexes before the last interpolated string
+                // handler. We _could_ limit this further by only spilling declaration expressions if the handler in question actually
+                // referenced a specific declaration expression in the argument list, but we think that the complexity in implementing
+                // this check is more complexity than this scenario needs.
```
>complexity [](http://example.com/codeflow?start=38&length=10)

"complexity" used twice. Consider rephrasing.

AlekseyTs reviewed (approved):
LGTM (commit 17), assuming CI is passing.

333fred commented:
@dotnet/roslyn-compiler @RikkiGibson for a second review.
